Add spec for tracing Azure Functions (#716)
elastic/apm: specs/agents/tracing-instrumentation-azure-functions.md

Depending on the actual kind of invocation (trigger type) there might be an additional object to retrieve metadata and context from.
In case of an HTTP invocation/trigger there is e.g. [`HttpRequestData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.httprequestdata).

## Generic Instrumentation

In general, to instrument an Azure Functions application, we create transactions that wrap the execution of its handler methods. In cases where we cannot assume any trigger type or extract trigger-specific data (e.g. if the trigger type is unsupported),
we wrap the handler method with a transaction, while using the generic context object to derive necessary fields:

| Field | Value | Description | Source |
| - | - | - | - |
| `name` | e.g. `MySampleTrigger` | The transaction name. Use function name if trigger type is `other`. | *generic context* |
| `type` | e.g. `request`, `messaging` | The transaction type. | Use `request` if trigger type is undefined. |
| `outcome` | `success`/`failure` | Set to `failure` if a function error can be detected otherwise `success`. | |

> | `result` | `success`/`failure` |Set to `failure` if a function error can be detected, otherwise `success`. | Trigger specific. |

Author: Agreed, I will add a note about that.

| `faas.name` | e.g. `MySampleTrigger` | The function name. | *generic context* |

> | `faas.id` | e.g. `/subscriptions/d2ba53be-0815-4...` | The [fully qualified resource ID](https://learn.microsoft.com/en-us/rest/api/resources/resources/get-by-id) of the Azure Function, which has this format: `/subscriptions/<SUBSCRIPTION_GUID>/resourceGroups/<RG>/providers/Microsoft.Web/sites/<FUNCAPP>/functions/<FUNC>` | *generic context*, environment |

Author: Correct, some of these environment variables (region, subscription, ...) are not available in a local `func` requirement.
I experimenting with an actual Azure Function that returns its whole environment as part of the response.

| `faas.trigger.type` | `other` | The trigger type. Use `other` if the trigger type is unknown or cannot be specified. | More concrete triggers are `http`, `pubsub`, `datasource`, `timer` (see specific triggers below). |

> | `faas.execution` | `203621a2-62f...` | The unique invocation id of the function. | *generic context* |

Author: Yes (in .NET it is `FunctionContext.InvocationId`).

> | `faas.coldstart` | `true`/`false` | A boolean value indicating whether this function invocation was a cold start or not. | [see section below](deriving-cold-starts)

Author: Yes, see https://github.com/elastic/apm/pull/712

> | `faas.coldstart` | `true`/`false` | A boolean value indicating whether this function invocation was a cold start or not. | [see section below](deriving-cold-starts)

Reviewer: I may be wrong, but #712 is about `metadata.cloud.provider` and this is about `transaction.context.cloud.origin.provider`. Two separate (and redundant?) fields.

> | `faas.coldstart` | `true`/`false` | A boolean value indicating whether this function invocation was a cold start or not. | [see section below](deriving-cold-starts)

Author: Right, I should have given more detail in my answer above.
Yes, it's a different field but I left it out on purpose, because:
* It's redundant.
* The way it was explained to me, all the `transaction.context.cloud.origin.*` fields represent the invocation side and cover a specific Lambda use case.

> | `faas.coldstart` | `true`/`false` | A boolean value indicating whether this function invocation was a cold start or not. | [see section below](deriving-cold-starts)

Reviewer: Makes sense. Thanks for the details.

### Metadata

Generic cloud metadata fetching SHOULD be disabled when the agent is running in an Azure Functions app. It will not
yield any useful data and only incur additional cold start overhead due to the non-existent HTTP metadata endpoints.

Agents SHOULD rather:

* Check for the existence of the `FUNCTIONS_WORKER_RUNTIME` environment variable that indicates that the agent is running
in an Azure Function application.
* Perform specific cloud metadata detection for Azure Functions (see [Metadata](./metadata.md)).

In addition to those described in [Metadata](./metadata.md), following metadata fields are relevant for Azure Functions:

| Field | Value | Description | Source |

> | - | - | - | - |

Author: I did not yet add entries for `service.name` and `service.version` like we have for [AWS Lambda](https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-aws-lambda.md#overwriting-metadata). Currently, those will be set like for any other application.
I recall that you had an opinion about that @AlexanderWert - using an Azure Functions-specific setting here?

Author: Definitely makes sense. In fact, that's what happens implicitly for .NET Azure Functions right now.
The assembly/application name is the `FUNCAPP` is the `service.name`.
We should however state this explicitly if this is the behavior we want. @gregkalapos @AlexanderWert

Reviewer: We could use the `WEBSITE_SITE_NAME` envvar as is for the `service.name`.

| `service.framework.name` | `Azure Functions` | Constant value for the framework name. | |
| `service.runtime.name`| e.g. `dotnet-isolated` | The language worker runtime (see [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#functions_worker_runtime)). | `FUNCTIONS_WORKER_RUNTIME` |

> ### Deriving cold starts

Reviewer: > Does that vary by language runtime?

https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=in-process%2Cfunctionsv2&pivots=programming-language-csharp#example suggests that for C# both in-process and out-of-process exists.

> ### Deriving cold starts

Author: > Does that vary by language runtime?

I assume it COULD vary by the language runtime and Microsoft will probably not lock itself in by promising it will be this or that way forever.
However, from what I observed so far (.NET, Node.js, Python): if a function app has multiple function entry points, these can run concurrently in the same process.

Ad ".NET 's `in-process` and `out-of-process` mode": This is not really related to that. Both of those hosting models
will allow concurrent function invocations in the same process.

Hence, cold start detection must happen in a thread-safe manner.

**Note:** The Azure Functions [Premium plan](https://learn.microsoft.com/en-us/azure/azure-functions/functions-scale)
guarantees pre-warmed workers (no cold starts). Nonetheless, reporting the first function execution as such might still make
sense since it could come with additional processing due to function-specific initialization overhead or JIT costs.  
Alternatively, agents could try to detect their hosting plan and not report cold starts for Premium plans at all.

### Disabled Functionality

Agents SHOULD detect whether they are running in an Azure Functions environment by testing
for the existence of the `FUNCTIONS_WORKER_RUNTIME` environment variable.

The following agent functionalities SHOULD to be turned off when tracing Azure Functions until decided otherwise:

* **Metrics collection:** this includes all kind of metrics: system, process and breakdown metrics and is equivalent to
setting `ELASTIC_APM_METRICS_INTERVAL = 0`
* **Remote configuration:** equivalent to setting `ELASTIC_APM_CENTRAL_CONFIG = false`
* **Cloud metadata discovery:**

>   * Equivalent to setting `ELASTIC_APM_CLOUD_PROVIDER = none`

Reviewer 2: I think this part clashes with the general idea of the `ELASTIC_APM_CLOUD_PROVIDER` config.

The point of that config is that the user just puts their cloud provider name there (in this case `azure`) and from that point the agent will only try to collect metadata for that provider.

This is maybe implementation detail, but if I remember correctly, setting this to `none` will also disable `AzureFunctionsMetadataProvider` in the .NET agent currently - meaning setting it to `none` will in general disable all cloud metadata collection - and this, in the current code includes the new one specifically for Azure Functions suggested in the line below. And this behaviour reflects the way this config is intended to work.

I'd expect other agents to work similarly.

The point of the spec here is clear (only use the new Azure Functions specific metadata collection suggested in the next line), I'm just not sure the `ELASTIC_APM_CLOUD_PROVIDER = none` reference is ideal.

Author: Could you take another look here @gregkalapos? Thanks

Reviewer 2: I guess I understand the intention and it totally makes sense to me: let's not try to query AWS, Google, and in Azure non Azure Functions related endpoints.

> or - that's what I had in mind - turn off the generic mechanism and just run the Azure Functions specific code since we know we are running as such due the presence of certain environment variables.

Yes, I also think this is the way to go, but still I'd not make this equivalent to `ELASTIC_APM_CLOUD_PROVIDER = none`, because in https://github.com/elastic/apm/pull/712/ that'll be listed as any other metadata collection. So setting it to `none` would make me think that none of the metadata collection listed in that list will run.

This is what [the spec says](https://github.com/elastic/apm/blob/main/specs/agents/metadata.md#cloud-provider-metadata):
> This metadata collection is controlled by a configuration value, CLOUD_PROVIDER. ... 
If set to none, no cloud metadata will be generated.

And that's not the case here, right? We still do generate cloud metadata - or don't we?

I just looked at the .NET code and this part is not yet implemented there. 

Would it make sense for quickly hop on zoom and discuss? I feel we agree on what we want here, but still this `ELASTIC_APM_CLOUD_PROVIDER = none` part sounds a bit strange to me when I see the rest of the spec.

Author: Yes, I think we agree on the general idea 👍 .

I'm not married to the exact wording here.
In fact, I took this from the [AWS Lambda spec](https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-aws-lambda.md). So we might need to reconsider the wording around `ELASTIC_APM_CLOUD_PROVIDER = none` there as well, I guess.

> I just looked at the .NET code and this part is not yet implemented there.
That's right.

Reviewer 2: Discussed over zoom, outcome: we changed the wording here.

However, depending on the trigger type, different information might be available that can be used
to capture additional transaction data or that allows additional, valuable spans to be derived.

### HTTP Trigger

HTTP invocations typically provide *some* variants of an HTTP *request object* (e.g. [`HttpRequestData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.httprequestdata))
and an HTTP *response object* (e.g. [`HttpResponseData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.functions.worker.http.httpresponsedata)).

Agents SHOULD use the information in the *request* and *response objects* to
fill the HTTP context (`context.request` and `context.response`) fields in the same way it is done for HTTP transactions.

In particular, agents MUST use HTTP headers to retrieve the `traceparent` and the `tracestate`
and use those to start the transaction for the tracing the function execution.

In addition the following fields should be set for HTTP trigger invocations:

| Field | Value | Description | Source |
| - | - | - | - |
| `type` | `request`| Transaction type. Constant value for HTTP trigger invocations. | |

> | `name` | e.g. `GET /prod/proxy/{proxy+}` | Transaction name: HTTP method followed by a whitespace and the (resource) path. | *request object* |

Reviewer: This example is from a Lambda triggered by API Gateway. If there is some real example that would be better to use.

The "req" object I see in a (locally run with `func`) JavaScript Azure Function with an HTTP trigger includes these fields. 

```
[2022-11-16T00:08:16.333Z]   req: {
[2022-11-16T00:08:16.334Z]     method: 'GET',
[2022-11-16T00:08:16.334Z]     url: 'http://localhost:7071/api/HttpExample?name=asdf',
[2022-11-16T00:08:16.334Z]     originalUrl: 'http://localhost:7071/api/HttpExample?name=asdf',
```

There is also some potential route pattern involved. E.g. https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=in-process%2Cfunctionsv2&pivots=programming-language-javascript#customize-the-http-endpoint shows an example HTTP binding configuration with URL parameters:

```json
{
    "bindings": [
    {
        "type": "httpTrigger",
        "name": "req",
        "direction": "in",
        "methods": [ "get" ],
        "route": "products/{category:alpha}/{id:int?}"
    },
```

In this case, I think we'd want the `transaction.name` to be "GET /.../projects/{category:alpha}/{id:int?}"  (where I'm not sure what, if anything, "/.../" should be.

Author: > This example is from a Lambda triggered by API Gateway. If there is some real example that would be better to use

Sorry, I copied this part over from the Lamba spec and obviously forgot to update the sample value here 🤦 .

> There is also some potential route pattern involved ...

That's valuable input. I'll add this to the spec.

However, depending on the trigger type, different information might be available that can be used
to capture additional transaction data or that allows additional, valuable spans to be derived.

### HTTP Trigger

HTTP invocations typically provide *some* variants of an HTTP *request object* (e.g. [`HttpRequestData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.httprequestdata))
and an HTTP *response object* (e.g. [`HttpResponseData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.functions.worker.http.httpresponsedata)).

Agents SHOULD use the information in the *request* and *response objects* to
fill the HTTP context (`context.request` and `context.response`) fields in the same way it is done for HTTP transactions.

In particular, agents MUST use HTTP headers to retrieve the `traceparent` and the `tracestate`
and use those to start the transaction for the tracing the function execution.

In addition the following fields should be set for HTTP trigger invocations:

| Field | Value | Description | Source |
| - | - | - | - |
| `type` | `request`| Transaction type. Constant value for HTTP trigger invocations. | |

> | `name` | e.g. `GET products/{category:alpha}/{id:int?}`, `GET /api/SampleHttpTrigger` | Transaction name: HTTP method followed by a whitespace and low cardinality representation of the function (route pattern or resource path). | *request object* |

Reviewer 3: Do we have any plans/ideas for how to get the low-cardinality name? `HttpRequestData` appears to only have the request uri, not any route information.

Author: Pragmatically, I would say this is best-effort depending on what's provided for the language/runtime in question.
As @trentm [pointed out](https://github.com/elastic/apm/pull/716#discussion_r1023405630), this information seems to be available in Node.js.

Reviewer: > are you able to get that route from within the runtime?

@basepi  I'm not sure that it is. My earlier comment was an observation of the config file from the developer side. I haven't yet played with real deployed Azure Fns.

Reviewer: (/cc @beniwohli)

I have some alternative suggestions for `transaction.name` for discussion. Sorry this got long.

An Azure Function app deployment looks like this (at least a Node.js one on windows
using [`WEBSITE_RUN_FROM_PACKAGE=1`](https://learn.microsoft.com/en-us/azure/azure-functions/run-functions-from-deployment-package)):

    C:\home\site\wwwroot\
        host.json           // Config for the function app
        HttpFn1\          // This dirname is the "function name"
            function.json   // Some config for the function, including a "route" template
            index.js        // The code for the function.
        HttpFn2\
            function.json   // Say has `"route": "products/{category:alpha}/{id:int?}"`
            index.js

Obviously "index.js" differs for other languages. My understanding is that "function.json" is a generated file for the compiled languages.

For HTTP-triggered functions, the URL to invoke it is:

    http(s)://<HOST:PORT><PREFIX>/<ROUTE>

The default PREFIX is `/api` and the default ROUTE is the function name.
For example:

    http://localhost:7071/api/HttpFn1     # locally using `func start`
    https://trentm-play-az-fn-app-1.azurewebsites.net/api/httpfn1

Notes ([Azure docs link](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=in-process%2Cfunctionsv2&pivots=programming-language-javascript#customize-the-http-endpoint)):

- The PREFIX can be configured (including to the empty string) via `extensions.http.routePrefix` in "host.json".
- The ROUTE can be configured in "function.json" (or in some code annotations, presumably).
- Also, note that the path (at least for *Windows*-deployed apps) is case-insensitive (yeah IIS), so we should avoid using the request URL path for the transaction name.

The currently spec'd examples:

> `GET products/{category:alpha}/{id:int?}`, `GET /api/SampleHttpTrigger`

are inconsistent. One is `<METHOD> <ROUTE>`, one is `<METHOD> <PREFIX>/<ROUTE>`.
I think we should consider one of the following.

## option A: `METHOD ROUTE`

Examples:

    GET HttpFn1
    GET products/{category:alpha}/{id:int?}

This would require instrumentation code to get the possible route template. The only way I found to get this was to read and parse the "function.json" on first invocation of a particular function, and then cache that. My [implementation is here](https://github.com/elastic/apm-agent-nodejs/blob/c0c2d0995e4edd6cd868872064ee9ccc29f9506e/lib/instrumentation/modules/azure-functions-nodejs-worker/dist/src/worker-bundle.js#L70-L102).

This code relies on the `context.executionContext.functionDirectory` field, which is perhaps notably *not* defined on the Java "ExecutionContext" interface: https://learn.microsoft.com/en-us/java/api/com.microsoft.azure.functions.executioncontext?view=azure-java-stable
*Perhaps* the path `C:\home\site\wwwroot\${executionContext.functionName}\function.json` could be assumed?

<img width="391" alt="Screen Shot 2022-12-19 at 5 00 22 PM" src="https://user-images.githubusercontent.com/46866/208556903-94349136-2105-40cc-a492-6b7bc6473c4e.png">

## option B: `METHOD PREFIX/ROUTE`

Examples:

    GET /api/HttpFn1
    GET /api/products/{category:alpha}/{id:int?}

This would require instrumentation code to also load the "host.json" file to get a possibly customized `routePrefix`. I haven't implemented this yet.

---

Thoughts?

Author: Right, I think having the full route is preferable here.
Especially in cases where developers change the default from `api` to something else, they would expect this name to be reflected in the span names.

(As for the mentioned inconsistency - this was not my intention. I meant for the full route to be present in both cases in the beginning).

## Generic Instrumentation

In general, to instrument an Azure Functions application, we create transactions that wrap the execution of its handler methods. In cases where we cannot assume any trigger type or extract trigger-specific data (e.g. if the trigger type is unsupported),
we wrap the handler method with a transaction, while using the generic context object to derive necessary fields.

**Note:** This table represents generic values that agents SHOULD always provide regardless of the
actual `faas.trigger.type`. For trigger types that are specifically supported (e.g HTTP) these values
can vary on trigger-specific specifications apply.

| Field | Value | Description | Source |
| - | - | - | - |
| `name` | e.g. `MySampleTrigger` | The transaction name. Use function name if trigger type is `other`. | *generic context* |
| `type` | e.g. `request`, `messaging` | The transaction type. | Use `request` if trigger type is undefined. |
| `outcome` | `success`/`failure` | Set to `failure` if a function error can be detected otherwise `success`. | |
| `result` | `success`/`failure` |Set to `failure` if a function error can be detected, otherwise `success`. | Trigger specific. |

> | `faas.name` | e.g. `MySampleTrigger` | The function name. | *generic context* |

Reviewer: [OTel semconv](https://opentelemetry.io/docs/reference/specification/resource/semantic_conventions/faas/#faas-resource-attributes) suggests `faas.name` should be `${functionAppName}/${functionName}`. What do you think about following that?  The app name is in the `WEBSITE_SITE_NAME` envvar. At least in JS, the function name is in `context.executionContext.functionName`, where `context` is the first arg to the handler function.

Author: Yes, good suggestion. Since we can have multiple function entry points in a function app, this is definitely more descriptive.
PR updated 👍 .

### Metadata

Generic cloud metadata fetching SHOULD be disabled when the agent is running in an Azure Functions app. It will not
yield any useful data and only incur additional cold start overhead due to the non-existent HTTP metadata endpoints.

Agents SHOULD rather:

* Check for the existence of the `FUNCTIONS_WORKER_RUNTIME` environment variable that indicates that the agent is running
in an Azure Function application.
* Perform specific cloud metadata detection for Azure Functions (see [Metadata](./metadata.md)).

In addition to those described in [Metadata](./metadata.md), following metadata fields are relevant for Azure Functions:

| Field | Value | Description | Source |
| - | - | - | - |
| `service.framework.name` | `Azure Functions` | Constant value for the framework name. | |
| `service.framework.version` | e.g. `~4` | Version of the Azure Functions runtime. | `FUNCTIONS_EXTENSION_VERSION` |
| `service.runtime.name`| e.g. `dotnet-isolated` | The language worker runtime (see [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#functions_worker_runtime)). | `FUNCTIONS_WORKER_RUNTIME` |

> 

Reviewer: ```suggestion
| `service.node.configured_name` | e.g. '25d4009bce1d ...' | Unique ID of the VM instance. | `WEBSITE_INSTANCE_ID` ([Azure docs](https://learn.microsoft.com/en-us/azure/app-service/reference-app-settings#scaling)) |
```

The `service.node.configured_name` is preferred over `system.hostname` for the entries in the "Instances" table on the service overview page, e.g.:
<img width="1065" alt="Screen Shot 2022-12-15 at 2 15 25 PM" src="https://user-images.githubusercontent.com/46866/207978818-f5d50633-5a2c-41a3-bbc1-565c82bd944c.png">
(The '8c1ff...' is from `WEBSITE_INSTANCE_ID`; the "10-20-0-211" is the hostname already sent in metadata at `system.hostname`.)

Reviewer 4: For Lambda, we disabled the instances view in the UI. Should we do the same for Azure functions?

Reviewer: @AlexanderWert Oh interesting. IIUC we have "Active Instances" on the Metrics page for Lambda (per https://github.com/elastic/kibana/issues/142328). If we had an equivalent for Azure functions, then yes, the "Instances" view would be somewhat redundant.

Azure Functions docs feel less concrete than AWS Lambda docs on the lifetime of the instances. The Azure Function VMs are *re-used* for deployment updates of function apps, so they are somewhat less ephemeral than Lambda instances.  Also, given that a single Azure Function instance can be concurrently handling requests for separate Functions in the same Function App, and hence they can have impact on each other (share CPU/memory), there is perhaps a *little* more of a basis for showing possible correlation of perf issues.

Author: As discussed yesterday @trentm, let's add the property to the spec.
We can then decide later whether we want to leverage that data in the UI.

> 

Reviewer: ```suggestion
| `cloud.provider` | `azure` | - | - |
| `cloud.region` | e.g. 'West US 2' | The cloud region. | `REGION_NAME` |
| `cloud.service.name` | 'azure functions' | Constant for the Azure service. See [ECS fields doc](https://www.elastic.co/guide/en/ecs/current/ecs-cloud.html#field-cloud-service-name). | - |
| `cloud.account.id` | e.g. '25b4db88-...' | The cloud id of the function. | Parsed from `WEBSITE_OWNER` if it is defined. `WEBSITE_OWNER_NAME && WEBSITE_OWNER_NAME.split('+', 1)[0]`
```

ECS documents `cloud.service.name` field here: https://www.elastic.co/guide/en/ecs/current/ecs-cloud.html#field-cloud-service-name Our Lambda spec uses "lambda" as a constant for this. Perhaps "functions" (feels too generic) or "azure functions"?

These `metadata.cloud` fields are responsible for the cloud icon next to the service name, e.g.:
<img width="421" alt="Screen Shot 2022-12-15 at 2 21 35 PM" src="https://user-images.githubusercontent.com/46866/207981386-ee0c2804-148e-4f7d-8d35-b303cad9df9a.png">

The `WEBSITE_OWNER_NAME` UUID value matches the value from `az account list`:
```
% az account list
[
  {
    "cloudName": "AzureCloud",
    "homeTenantId": "39f4034f-9fa9-4141-969e-94a6032fad22",
    "id": "25b4db88-9548-4a22-955f-31b826b06078",
    "isDefault": true,
    "managedByTenants": [],
    "name": "[REDACTED]",
    "state": "Enabled",
    ...
```

Author: Right, I must have missed this property initially. No intention of not sending it.
As for the name, no strong opinion. So, let's go with `"functions"` (for now).

However, depending on the trigger type, different information might be available that can be used
to capture additional transaction data or that allows additional, valuable spans to be derived.

### HTTP Trigger

HTTP invocations typically provide *some* variants of an HTTP *request object* (e.g. [`HttpRequestData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.httprequestdata))
and an HTTP *response object* (e.g. [`HttpResponseData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.functions.worker.http.httpresponsedata)).

Agents SHOULD use the information in the *request* and *response objects* to
fill the HTTP context (`context.request` and `context.response`) fields in the same way it is done for HTTP transactions.

In particular, agents MUST use HTTP headers to retrieve the `traceparent` and the `tracestate`
and use those to start the transaction for the tracing the function execution.

In addition the following fields should be set for HTTP trigger invocations:

| Field | Value | Description | Source |
| - | - | - | - |
| `type` | `request`| Transaction type. Constant value for HTTP trigger invocations. | |

> | `name` | e.g. `GET /api/MyFuncName`, `GET /api/products/{category:alpha}/{id:int?}` | `<HTTP-method> /<route-prefix>/<route-pattern-or-function-name>` | `<HTTP-method>` from the request object. `<route-prefix>` from "extensions.http.routePrefix" in host.json, defaults to `api`. `<route-pattern>` from "[`<function-dir>`](https://learn.microsoft.com/en-us/azure/azure-functions/functions-referencefolder-structure)/function.json", defaults to the function name. [Azure docs link](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger#customize-the-http-endpoint). |

Reviewer: ```suggestion
| `name` | e.g. `GET /api/MyFuncName`, `GET /api/products/{category:alpha}/{id:int?}` | `<HTTP-method> /<route-prefix>/<route-pattern-or-function-name>` | `<HTTP-method>` from the request object. `<route-prefix>` from "extensions.http.routePrefix" in host.json, defaults to `api`. `<route-pattern>` from "[`<function-dir>`](https://learn.microsoft.com/en-us/azure/azure-functions/functions-referencefolder-structure)/function.json", defaults to the function name. The route must be normalized to have exactly one leading slash. [Azure docs link](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger#customize-the-http-endpoint). |
```

In addition to those described in [Metadata](./metadata.md), following metadata fields are relevant for Azure Functions:

| Field | Value | Description | Source |
| - | - | - | - |
| `service.name` | e.g. `MyFunctionApp` | If the service name is *explicitly* specified through the `service_name` agent config option, use that value. Otherwise, use the name of the Function App. | If `service_name` is not specified, use `WEBSITE_SITE_NAME`. |
| `service.name` | e.g. `MyFunctionApp` | If the service name is *explicitly* specified through the `service_name` agent config option, use that value. Otherwise, use the name of the Function App. | If `service_name` is not specified, use `WEBSITE_SITE_NAME`. |
| `service.framework.name` | `Azure Functions` | Constant value for the framework name. | |
| `service.framework.version` | e.g. `~4` | Version of the Azure Functions runtime. | `FUNCTIONS_EXTENSION_VERSION` |
| `service.runtime.name`| e.g. `dotnet-isolated` | The language worker runtime (see [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#functions_worker_runtime)). | `FUNCTIONS_WORKER_RUNTIME` |

> | `service.node.configured_name` | e.g. `25d4009bce1d ...` | Unique ID of the VM instance. | `WEBSITE_INSTANCE_ID` ([Azure docs](https://learn.microsoft.com/en-us/azure/app-service/reference-app-settings#scaling)) |

Author: Hi @trentm,
@gregkalapos [made me aware of](https://github.com/elastic/apm-agent-dotnet/pull/1959#discussion_r1055837164) the fact that this property can be set via [configuration](https://www.elastic.co/guide/en/apm/agent/dotnet/current/config-core.html#config-service-node-name).

So, we could:
* Take take the same approach as with `service.name` and only use the environment variable if no config option is present. However, I don't think the config option makes a lot of sense in an Azure Functions context.
* Use `WEBSITE_INSTANCE_ID` in any case and maybe print a log message if this would override the config option.

My preferred option would be the latter. WDYT?

Reviewer: @z1c0  @gregkalapos 

I looked back at the history of `metadata.service.node.configured_name`.
It was added in APM Server 7.5 (released December 2, 2019). Discussion at: https://github.com/elastic/apm/issues/141
In [this comment](https://github.com/elastic/apm/issues/141#issuecomment-530280227) the intake field name `configured_name` was suggested -- to be similar to the already existing `metadata.system.configured_hostname` and `metadata.system.detected_hostname` -- saying:

> This allows us to add a `service.node.[detected]_name` in the future.

I think Azure Functions and AWS Lambda are reasonable cases for a new `metadata.service.node.detected_name`. However, that would be separate work, and would obviously then require a newer APM Server version. So I'll ignore that idea for now.

* * *

I agree that using the `service_node_name` config var doesn't make sense for a serverless system like Azure Functions.  My (somewhat unsure) understanding from the above issue is that the intent of `service_node_name` is for multiple instances of the same service running on the same node and that setting it could result in some harm -- from [this comment](https://github.com/elastic/apm/issues/141#issuecomment-529822151)]:

> I have the fear that making the name configurable could to more harm than good. We heavily rely on the value being unique across a cluster.

Therefore:

- I think `WEBSITE_INSTANCE_ID`, if it exists, should "win" over a configured `service_node_name`.
- Possibly the `service_node_name` config var should explicitly be ignored in an Azure Function or Lambda. (If we think so, this should be a separate spec PR. If not, then at the least, agent documentation for the `service_node_name` config var should say to not use it in serverless configs.)

Thoughts?

Agents SHOULD rather:

* Check for the existence of the `FUNCTIONS_WORKER_RUNTIME` environment variable that indicates that the agent is running
in an Azure Function application.
* Perform specific cloud metadata detection for Azure Functions (see [Metadata](./metadata.md)).

In addition to those described in [Metadata](./metadata.md), following metadata fields are relevant for Azure Functions:

| Field | Value | Description | Source |
| - | - | - | - |
| `service.name` | e.g. `MyFunctionApp` | If the service name is *explicitly* specified through the `service_name` agent config option, use that value. Otherwise, use the name of the Function App. | If `service_name` is not specified, use `WEBSITE_SITE_NAME`. |
| `service.framework.name` | `Azure Functions` | Constant value for the framework name. | |
| `service.framework.version` | e.g. `~4` | Version of the Azure Functions runtime. | `FUNCTIONS_EXTENSION_VERSION` |
| `service.runtime.name`| e.g. `dotnet-isolated` | The language worker runtime (see [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#functions_worker_runtime)). | `FUNCTIONS_WORKER_RUNTIME` |

> | `service.node.configured_name` | e.g. `25d4009bce1d ...` [1] | Unique ID of the VM instance. | `WEBSITE_INSTANCE_ID` ([Azure docs](https://learn.microsoft.com/en-us/azure/app-service/reference-app-settings#scaling)) |
> 
> [1] **Note** that `service.node.configured_name` can also be set via the agent `service_node_name` configuration. In that case, the Azure Functions
> instrumentation MUST override the existing value and MUST print a log message accordingly.

Reviewer: ```suggestion
| `service.node.configured_name` | e.g. `25d4009bce1d ...` | Unique ID of the VM instance. | `WEBSITE_INSTANCE_ID` ([Azure docs](https://learn.microsoft.com/en-us/azure/app-service/reference-app-settings#scaling)) **Note:** If the `service_node_name` agent configuration setting is provided (which normally will set `service.node.configured_name`) it MUST be ignored in favor of `WEBSITE_INSTANCE_ID`, and a log warning MUST be emitted saying it was ignored. |
```

@z1c0 How about this wording?

Author: I like it, much clearer. Thanks @trentm!
